### ISSUE TITLE
Structure and naming of new commands

### ISSUE THREAD

**Issue by robinkrahl** [d-e-s-o/nitrocli#4]
The current command handling code has two limitations:
- It does not support arguments.
- It only supports command names that are valid function names in Rust.

There are some good argument parsing libraries for Rust, but they are rather heavy-weight.

I think we currently need these commands (grouped by topic):

## General
- [x] `status`: return general and model-specific information (firmware version, serial number, retry counts, storage status for NK Storage)
- [x] `change-pin`: change user or admin PIN
- [x] `unlock-pin`: unlock the user PIN
- [x] `lock`: lock the device (i. e. password store and volume, if applicable)

## Configuration (`config`)
- [x] `get`: read the current configuration
- [x] `set`: change the configuration

## OTP (`otp`)
- [x] `get`: generate a OTP
- [x] `set`: configure a OTP slot
- [x] `clear`: clear a OTP slot
- [x] `status`: display all slots

## Password Safe (`pws`)
- [x] `status`: display the status of all alots
- [x] `get`: get the data stored on one slot
- [x] `set`: configure a slot
- [x] `clear`: clear a slot

## Storage (`storage` or `volume`)
- [x] `open`: enable encrypted volume
- [x] `close`: disable encrypted volume

So my questions are:
- How should we name the commands?  We could use names like `get-otp` or `otp-get`.  But I think the most intuitive solution would be to have subcommands (like `opt [get|set]` or `pws [status|get|set|clear]`).
- How should we parse the arguments?  Should we implement it from scratch or use one of the available libraries?

**Comment by d-e-s-o** [comment]
Thanks for this summary of what you see needed.

> * How should we name the commands?  We could use names like `get-otp` or `otp-get`.  But I think the most intuitive solution would be to have subcommands (like `opt [get|set]` or `pws [status|get|set|clear]`).

Yes, I am very much in favor of having subcommands. I found those to be the most intuitive and easy to type/remember/navigate.

> * How should we parse the arguments?  Should we implement it from scratch or use one of the available libraries?

That is a good question. I am not big on reimplementing this logic, as it is error prone and typically leads to bad user experience if not done perfectly. Plus it's wasted effort. If all we need is subcommands that may not be a big deal, but we will very likely also need additional parameters for those, and then it gets tricky.

Back in the day I did investigate using [`clap`](https://crates.io/crates/clap) (either for this or a similarly small crate), but I found the bloat it caused to be unbearable for such a small application. If I remember correctly the size of the binary did double or something like that. And that was after most of the work for size reduction on their side had already landed.
We may want to reinvestigate quickly if something changed significantly, but if not it may be overblown for our use case.
A second argument I usually take into account is the number of dependencies (transitive and direct). That should be as low as possible.
Lastly, given the subcommand requirement, the crate we use likely needs to be capable of handling those natively.
Oh, and it should also be still maintained ;)

Other than `clap` a quick search yielded:
* [`structopt`](https://crates.io/crates/structopt): based on `clap` so the same main argument applies
* [`docopt`](https://crates.io/crates/docopt): Slightly smaller set of direct dependencies compared to `clap`. It seems to have full sub-command support from a brief look of it. I am not familiar with Docopt in general, but I may read up on it some more.
* [`rust-argparse`](https://github.com/tailhook/rust-argparse): No dependencies is a plus. Seems to be similar to Python's argparse which supports subcommands.
* [`getopts`](https://crates.io/crates/getopts): Low number of dependencies. But I doubt it has subcommand support.

(Did I miss something?)

Off hand I'd be very much leaning towards giving `rust-argparse` a try. I have not yet played around with it or read the documentation more closely, so I can't tell whether it's a good fit.
But these are my thoughts.

**Comment by robinkrahl** [comment]
Thanks for the quick feedback. I agree with your evaluation of parsing libraries. I’ll implement the `opt set` command using `rust-argparse`, then we can evaluate it.

**Comment by robinkrahl** [comment]
I started to use `argparse` on my [`wip/argparse`][] branch (see 003c68fad74242305c26fcc47ae5bce8bed73f2f for the actual implementation).  I think `argparse` meets our needs, but there it requires some boilerplate code.  Note that my current implementation only has one layer of commands, so it will get even more verbose for the subcommands.  It is especially annoying that there are so many places where you have to define the mapping between command name, enum variant and the function to call.  I’ll try to write a module that makes it a bit easier to maintain the commands and to execute the parser. Any suggestions are welcome.

[`wip/argparse`]: https://github.com/robinkrahl/nitrocli/tree/wip/argparse

**Comment by robinkrahl** [comment]
I implemented `otp get` and `otp set` using `argparse` (see wip/argparse).  I think we can live with this solution.  We won’t be adding many commands after the initial implementation, and most of the functions will not compile if we forget to add new commands.  One option to make maintenance easier would be to use the `strum` crate to automatically implement `FromStr` and `Display` for `Command` and the subcommand enums based on attributes.

Also, I suggest to restructure the code as follows:
- `options.rs` for argument handling
- `commands.rs` for theapplication logic

So for example, we would have `options::otp_get(args: Vec<String>)` that parses the arguments and then calls `commands:otp_get(slot: u32, ...)`.

Another issue to discuss is the design of subcommands.  Should we use positional arguments or options for mandatory arguments?  For example, `otp set` always needs the slot number, the name of the slot and the secret to store on the slot.  The current implementation is:
```
nitrocli otp set 1 name secret --algorithm hotp --counter 10
```
While this is more concise and intuitive, it might be less confusing and error-prone to write:
```
nitrocli otp set --slot 1 --name name --secret secret --algorithm hotp --counter 10
```

**Comment by robinkrahl** [comment]
> Sounds as if a macro may be enough. But I guess I may be missing something. I'll try to have a look at the crate some time soon.

I think it would require a procedural macro, and that’s basically what that crate provides.  But so far I mostly avoided writing macros, so I might well be wrong about that.

> I think I'd go with making mandatory arguments positional. I am always confused when an application asks me to supply an option or seizes to work otherwise.

Okay. Then I’ll prepare a pull request for the commands.

**Comment by d-e-s-o** [comment]
Hey @robinkrahl , do you already have patches for the subcommand introduction of `volume open/close` (mentioned in the first comment) in the pipeline? Otherwise I can take care of those.

**Comment by robinkrahl** [comment]
No, all the patches I had prepared are the open pull requests.  I’ll deal with the PWS next, so it would be great if you could have a look at the volume refactoring.

**Comment by robinkrahl** [comment]
Speaking of the `volume` command, do you think it would make sense to move the Storage-specific status information from `status` to `volume status`?  That would be more consistent with `otp status` and `pws status`.

**Comment by d-e-s-o** [comment]
> No, all the patches I had prepared are the open pull requests. I’ll deal with the PWS next, so it would be great if you could have a look at the volume refactoring.

Sounds good!

> Speaking of the `volume` command, do you think it would make sense to move the Storage-specific status information from `status` to `volume status`? That would be more consistent with `otp status` and `pws status`.

That seems like the sensible thing to do. I'll look into that as well. Thanks for the suggestions!

**Comment by robinkrahl** [comment]
I’m about to implement the PIN unblock and change commands, yet I’m considering to drop them because they are already provided by `gpg --change-pin`.  What do you think?

**Comment by robinkrahl** [comment]
No, it is not much work to implement the functions.  The only issue is that, at least for my Nitrokey Storage, the un(b)lock command does not work as documented – you cannot reset the user PIN to arbitrary values, but you have to use the current user PIN.  I don’t know if this is a bug or a misconception on some side.  I reported the problem to the gnupg-users mailing list and will post a link here once the thread appears in the archive.

**Comment by robinkrahl** [comment]
https://lists.gnupg.org/pipermail/gnupg-users/2018-December/061411.html

**Comment by d-e-s-o** [comment]
Oh, interesting. Thanks for checking on that. Would it be possible to do the same with `libnitrokey` to see whether it's a `gpg` problem or more on the Nitrokey side (I would hope something like this is part of its API)?

**Comment by robinkrahl** [comment]
Yes, `nitrokey::unlock_user_pin`/`NK_unlock_user_password` has the same problem.  So it’s most likely an implementation issue on the smart card.

**Comment by robinkrahl** [comment]
As all suggested commands have been implementend, I’m closing this issue.

**Comment by szszszsz** [comment]
Hi! Regarding the unlocking issue, since I cannot access the answer on the GnuPG list. In case of locked User PIN, two options are possible:
- Unlocking User PIN, which requires Admin PIN, via either GnuPG or `nitrokey::unlock_user_pin/NK_unlock_user_password`. This operation will not clear data from the smart card.
- Resetting the smart card completely. This sets the standard PINs (both Admin and User), and clears all user data. See details [here](https://www.nitrokey.com/documentation/frequently-asked-questions-faq#how-to-reset-a-nitrokey).
